### DISCUSSION TITLE
Problem configuring hy-mode.el and jedhy

### DISCUSSION THREAD

**Discussion by prairie-guy** [hylang/hy#2055] [Q&A] [answered]
Long time user of Clojure, but never been a huge fan of it being coupled with Java. Imagine my excitement to discover Clojure built on top of Python as HY!

I’m trying Hy with emacs but am having trouble getting it working smoothly. Here is what I have done so far. (Linux 20.04, Anaconda python, conda 3.8)

1) pip3 install --pre --user hy
2) pip 3 install jedhy
3) Installed hy-mode.el within emacs (Doom, emacs 27.1)

Using the hy repl  from the terminal works as expected.
Within emacs, when I attempt the command `run-jedhy` I get error: `Jedhy failed to start`
When I use `run-hy`, an hy shell opens. 
```
import hy
hy 1.0a1 using CPython(default) 3.8.8 on Linux
=> 
```
I’ve made sure that `company-mode` is active.
Autocompletion does not appear to be working within the hy buffer. 
Moreover, within the hy shell, autocompletion returns the following (on a specific function name):

```
Traceback most recent call last:?  File "stdin-af759d2873eafdd40be18f511a7dcd87ac3f24f0" ... Traceback (most recent call last):?  File "stdin-17d6a  
 in <module>?    --JEDHY.complete "reverse"?NameError: name hyx_XhyphenHminusX_JEDHY is not defined ... ... Traceback (most recent call last):?  Fil  
 line 1 Traceback (most recent call last):?  File "stdin-76c0062a8a60a4f35093c32f6bf17b4a39da7832", line 1, in <module>?    (--JEDHY.annotate "line 
```

Other things within the hy buffer and shell are not working as expected.
I’ve clearly not configured this properly. Any advice appreciated.

Thanks in advance.

**Comment by allison-casey** [comment, marked as answer]
Hello! Yes this is a known issue. The pace of Hy's development is currently outpacing the ability of the supporting libs to keep up (we are a small dev team and the focus is stabilizing the language right now).  If you'd like to use Jedhy the current stable release of Hy `0.20.0` is the most widely supported version (which is what you'd get if you did a normal `pip install hy`) We're currently on an alpha cycle and once Hy has settled down and we've ironed out the last breaking changes we'll pivot to bringing everything up to date. If you'd like to use the latest alpha releases of Hy you will need to turn Jedhy off in hy-mode.el with `(setq hy-jedhy--enable? nil)`. Hopefully that answers you're question! let me know if you'd like more clarification on anything

**Reply by prairie-guy** [reply]
Allie - Many thanks! Yes that is a big help. Now it’s definitely usable. Makes a ton of sense. So glad there is continuing development on Hy. Looking forward to exploring Hy now that I can use it with emacs.

OK, I’m using Hy 0.20.0 and have been able to execute: `run-jedhy` But am still not getting autocompletion.

**Reply by allison-casey** [reply]
i'm not sure what you mean by `ry-repl` i assume you meant `hy-repl`.  if you're on `0.20` i believe you can get autocompletion, but there are limitations that you can read up on over at the jedhy [repo](https://github.com/ekaschalk/jedhy). mainly around the fact that jedhy isn't doing static analysis like its python sibling `jedi` is.

**Reply by prairie-guy** [reply]
Sorry, stubby fingers. Yes, `hy-repl`. I’m not getting autocompletion in either the `hy-buffer` or `hy-repl`. I’m not getting any of the company-mode stuff. (I do get highlighting and formatting.) You have been very generous with your time. Let me read through the repo as you suggest. As long as I know `company-mode` should be working and that I should be able to get autocompletion, let me see what I’m doing wrong. Best

**Comment by prairie-guy** [comment]
I’m trying to configure `hy-mode` such that `hy-repl` shell within in so that it acts like: 
`hy --repl-output-fn=hy.contrib.hy-repr.hy-repr`

Within my `config.el` file:
`(setq hy-shell--interpreter-args 'nil)` sets `repl-spy` to` False` as expected.
`(setq hy-shell--interpreter-args '("--repl-output-fn=hy.contrib.hy-repr.hy-repr”)` returns the following error on starting up error in `hy-repl`

Error: `hy: unrecognized option: --repl-output-fn\`

Is this an issue with mangling or use of delimiters?

Also, is there a default start-up file that hy uses or is the best way to set the environmental variable `HYSTARTUP`

If I am reading the documentation properly, it does look like I can set `repl-output-fn` at start-up. In a perfect world, I would like the set a global default start-up behavior, but still be able to override it within `config.el`.

Finally, is it possible to dynamically reset `hy-repl `behavior from within a running repl? I’ve tried executing “(setv repl-spy True)”

Appreciate the help. Looking to really start playing around with the language once I understand how to use it properly within emacs.

Thanks

**Reply by allison-casey** [reply]
you can set repl output fn with the following in your emacs configs 
`(setq hy-shell--interpreter-args '("--repl-output-fn" "hy.contrib.pprint.pformat"))` `hy-repr` or `pformat` are good choices here.

I don't believe you can change the `spy` behavior while the repl is running. Pretty sure that's a limitation of python's `codeop` classes that the repl is built on top of. You've made me realize though that if you have a HYSTARTUP file that sets the `spy` param it won't get overridden if yoiu manually pass --spy. i'll open a PR to fix that
